### DISCUSSION TITLE
How to deal with contextual and non serizalizable query key ?

### DISCUSSION THREAD

**Discussion by stevebeauge** [TanStack/query#6099] [Q&A]
I'm working on a project requiring some custom HTTP client to fetch API data. The project is actually a plugin of a 3rd party application that injects the plugin using a DOM node and a render method. The render method get a custom http client as parameter. There's no "global" object I can use to get such http client.
This works well, but it breaks react query dependency array check, because the class is not serializable (cyclic dependency).

I created a code sandbox repro to illustrate:

https://codesandbox.io/s/thirsty-shamir-p4y89x?file=/src/App.tsx

Both query client and custom Http client are provider to the react tree using context providers:

````typescript
export default function App() {
  return (
    <CustHttpClientProvider value={customHttpClient}>
      <QueryClientProvider client={queryClient}>
          .....
      </QueryClientProvider>
    </CustHttpClientProvider>
  );
}
````

This allows me to write component that requires both context objects:

````typescript
const UserInfoKO: React.FC<{ userId: number }> = ({ userId }) => {
  const httpClient = useCustHttpClient();
  const { data, status } = useQuery({
    queryKey: ["userInfo", userId, httpClient], // <=== My custom http client is a dependency
    queryFn: () => api(httpClient, userId)
  });

  return (
    <div>
      {status}: {data}
    </div>
  );
};
````

But this code fails because the httpClient variable cannot be serialized : `TypeError: cyclic object value`.

If I don't put this variable in the dependency array, the code runs but the eslint rule yell because of the missing dependency (this can be removed, but it may hide also other dependency miss.

I can trick react query using something like:

````
  const httpClient = useCustHttpClient();
  const apiWithClient = api.bind(null, httpClient); // <== the trick is there, 
  const { data, status } = useQuery({
    queryKey: ["userInfo", userId],
    queryFn: () => apiWithClient(userId)
  });
````
this works at runtime, eslint rule is doing its job because the argument is out of the query function.  But this requires a lot more non obvious code to add everytime an API is required.

Is there any proper way to handle such situation ?

Is there any way to "extend" the query client with arbitrary data ? Something like a custom "context" that may be set to query client ?

thanks for advise!

**Comment by stevebeauge** [comment]
Some explorations later, I have found the "meta" option of the useQuery hook

````typescript
const UserInfoOK3: React.FC<{ userId: number }> = ({ userId }) => {
  const httpClient = useCustHttpClient();
  const { data, status } = useQuery({
    queryKey: ["userInfo", userId],
    queryFn: ({ meta }: { meta: WithHttpClientMeta }) =>
      api(meta.httpClient, userId),
    meta: {
      httpClient
    }
  });

  return (
    <div>
      {status}: {data}
    </div>
  );
};
````

There's no, however, strong typing of the meta object, neither isolation.

**Reply by TkDodo** [reply]
In v5, you can type the meta function with module augmentation:

```
declare module '@tanstack/react-query' {
  interface Register {
    queryMeta: {
      httpClient?: HttpClient
    }
  }
}
```

However, I don't think that's what `meta` was made for. In v5, `meta` will also be part of hydration / serialization when using persister plugins.

We talked some time back about having an allowList for things that the lint rule should ignore - we also allow using `queryClient` from `useQueryClient()` directly in the `queryFn`. @Newbie012 this wasn't implemented, right?

The problem is: if `httpClient` changes reference, you might run into stale closure problems. That's not an issue with the queryClient, because if it changes, the cache gets wiped. So I'm not sure if we should allow to introduce exceptions.

The other solution is to implement a custom `queryKeyHashFn` globally that serializes the `httpClient`. Could just be a static value if it never changes, or, if there are things in there that could change (and should cause the queryFn to re-run), you could serialize those values only.

**Reply by Newbie012** [reply]
That's correct. There's currently no such option to ignore certain nodes in the query key. Every time it came up, there was a better solution.

**Reply by UNLRN** [reply]
@TkDodo @Newbie012 Hi, I'm running into a similar problem while trying to adopt `react-query` into an existing codebase. We pass a non-serializable utility object that has an api client (among other utilities) to an `api` function, almost identically to the original post.

```ts
  const utilities= useUtilities()
  const userRequest = useQuery({
    queryKey: [QUERY_KEY, userId], // exclude passing `utilities` to the queryKey
    queryFn: () => getUserById(userId, utilities),
  })
  ```

I've patched the eslint package in our project to allow ignoring this dependency. I'd happily add a PR for that, but it sounds like you're not open to adding this as a feature?
  
If that's the case:
  - Is the `queryKeyHashFn` the suggested way to deal with this?
  - If so, is there a way to globally add that logic or is it expected we'd have to do that on each usage of `useQuery`?
  - Are there other options that can be done globally?

**Reply by TkDodo** [reply]
the problem here is that `utilities` can change between renders. For example:

```
const useUtilities = () => {
  const [count, setCount] = useState(0)

  return [count, setCount]
}
```

now if you read `count` inside `getUserById`, you might read stale values if the count was updated.

So the question is: if `utilities` is stable, why does it have to be a hook / inside the react component at all? if it's not stable, you have to add whatever `getUserById` actually needs to key. If it's just the `count`, I'd make this explicit:

```
const utilities= useUtilities()
const userRequest = useQuery({
  queryKey: [QUERY_KEY, userId, utilities.count],
  queryFn: () => getUserById(userId, utilities.count),
})
```

of course count is a stupid example, but I hope you get what I mean

**Reply by UNLRN** [reply]
I understand the reasoning for requiring the dependencies in the `queryKey`. This is a codebase that already has the pattern of centralizing stable and non-stable utilities as a hook. I would have preferred (many times) to have the stable ones outside the react component, but it is what it is :) 

Per the docs:
> Since query keys uniquely describe the data they are fetching, they should include any variables you use in your query function that change.

In essence, these utilities don't describe in any way the data they are fetching. If the dynamic pieces update, we're not concerned with them impacting the query functions. Using the `queryKeyHashFn` is an alternative, but it doesn't seem to be applicable globally?

FWIW, I sympathize with wanting to keep the dependencies rule strict, as there's probably been a number of issues/questions that were solved by adding it in the first place and is ultimately why we're using the eslint rule too. I'd just like to find the best solution for when the particular value passed to the query function can be ignored. Happy to contribute if there is another path to enabling something like this globally, if it doesn't exist.

**Reply by TkDodo** [reply]
> Using the queryKeyHashFn is an alternative, but it doesn't seem to be applicable globally

you can pass it as a default for all your queries

> In essence, these utilities don't describe in any way the data they are fetching.

what do they do then - why are they needed in the queryFn? What I've often seen is that things like that are used to e.g. inject headers (tokens) and then stale closures will lead to sending invalid tokens.

So I'm not really sure how something can be passed to the queryFn that is stable but still comes from a hook and also doesn't affect the underlying `fetch` call at all. What would that be?

**Reply by stevebeauge** [reply]
I confirm @TkDodo's answer. Since my initial post, I've succeed in building react query production apps that requires some weird 3rd party objects in my async code.

My global query client is built like this:

````typescript
const customQueryClient = new QueryClient({
    defaultOptions: {
        queries: {
            refetchOnWindowFocus: false,
            retry: false,
            queryKeyHashFn: queryKeyHashFnCustom,
        },
    },
});
````

which defaults query key hash to a custom one. This function is similar to:

````typescript
const queryKeyHashFnCustom: QueryKeyHashFunction<QueryKey> = (key) => {
    const fixKeys = key.map((k) => {
        if (k instanceof WeirdType1) {
            return getNonSerializableUID(k);
        }
        if (k instanceof WeirdType2) {
            return k.instanceId;
        }
        if (hasQueryQueryKeyHashFn(k)) {
            return k.getQueryKeyHash();
        }
        if (queryKeyShouldUseWeakMap(k)) {
            return getNonSerializableUID(k);
        }
        return k;
    });
    return hashQueryKey(fixKeys);
};

// Store every non serializable object in a WeakMap, associated with a unique id.
const instanceMap = new WeakMap<object, string>();
function getNonSerializableUID(k: object): string {
    if (!instanceMap.has(k)) {
        const id = crypto.randomUUID();
        instanceMap.set(k, id);
    }

    return notnull(instanceMap.get(k));
}
````

Basically, the code will test if key is of type of some specific and non serializable classes, generate a random ID and store it inside a weak map (to avoid memory leak). This is useful for 3rd party classes I cannot extends.

I also test for some other type, also from 3rd party lib I cannot change, and use one of its property to get a uniqueID, and use this id to serialize the key.

Finally, for my own types, where internal implementation make the object non serializable, I define some interface I can implement to provide a custom key for the object. Here's the implementation of the typescript type guards:

````typescript
interface IHasQueryQueryKeyHashFn {
    getQueryKeyHash: () => string;
}

const hasQueryQueryKeyHashFn = (obj: unknown): obj is IHasQueryQueryKeyHashFn => {
    return (
        typeof obj === "object" &&
        obj !== null &&
        "getQueryKeyHash" in obj &&
        typeof (obj as IHasQueryQueryKeyHashFn).getQueryKeyHash === "function"
    );
};

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
interface IQueryKeyShouldUseWeakMap {
    useWeakMap: boolean;
}

const queryKeyShouldUseWeakMap = (obj: unknown): obj is IQueryKeyShouldUseWeakMap => {
    return (
        typeof obj === "object" &&
        obj !== null &&
        "useWeakMap" in obj &&
        typeof (obj as IQueryKeyShouldUseWeakMap).useWeakMap === "boolean"
    );
};
````
If I know a type won't work in the OOB query fn, I can implement either of the two interfaces, wheter I have some unique value, or if I need a week map:

```` typescript
abstract class BaseService implements IQueryKeyShouldUseWeakMap {
    protected readonly context: WeirdType;
    protected readonly webUrl: string;

    constructor(context: WeirdType, webUrl = context.pageContext.web.absoluteUrl) {
        this.context = context;
        this.webUrl = webUrl;
    }
    public readonly useWeakMap = true;
}
````

Ultimately, if none of the tests is satisfied, I let the react query native hash function handle the serialization.

**Reply by UNLRN** [reply]
> So I'm not really sure how something can be passed to the queryFn that is stable but still comes from a hook and also doesn't affect the underlying fetch call at all. What would that be?

There are utilities that are unrelated to fetching, like helpers that wrap libs/browser apis or just provide metadata of the current running env. But what goes on inside an API function that receives the utilities is sometimes more than just calling `utilities.fetch` so the whole `utilities` object gets passed instead of just `utilities.fetch` 😅

The ideal for me is to limit the surface area that would be impacted by migrating to react-query. Hiding some implementation details so that our developers can use `react-query` hooks as vanilla as possible without us adding our own abstractions over the hooks would be my preference, which is why I'm exploring options. If we can continue passing the whole `utilities` object and just serialize it properly, that's one option. But if we end up needing to pull items off `utilities` and pass them, there are some things that aren't relevant to the `queryKey` that we _could_ just ignore.

Regardless, we still have to handle serialization, so at least having the `queryKeyHashFn` at the global level will be useful.

@stevebeauge thanks for the code samples. It will be helpful to use this as a reference when we implement something similar.

**Reply by stevebeauge** [reply]
Side note : my code is for v4 (can't use react 18)
Maybe v5 has some change or improvement that can help

**Reply by TkDodo** [reply]
> There are utilities that are unrelated to fetching, like helpers that wrap libs/browser apis or just provide metadata of the current running env. But what goes on inside an API function that receives the utilities is sometimes more than just calling utilities.fetch so the whole utilities object gets passed instead of just utilities.fetch 😅

sure, but again, if they are stable, there is no need for them to be a hook that lives inside the react component tree. If it e.g. contains the BASE_URL env that is either localhost or your deployed url - why would that be a hook?

**Reply by jlowcs** [reply]
Adding some context over a year later:

Yes, `utilities` is stable, but across a render. For instance, client-side, it is created once at bootstrap and never modified, so it is stable, and could in theory be a global variable.

The issue is that we're using that `utilities` object on the server too (SSR). On the server, we can generate multiple pages in parallel for which `utilities` would *not* be the same. So, even though it is stable within a single render, it cannot be global, otherwise that would cause collisions. That is why it is retrieved through a hook even though it is considered stable client-side.

**Comment by stevebeauge** [comment]
Sorry for the long delay. To conclude, I followed your suggestion to implement queryKeyHashFn. I use this code:

````typescript
import { hashQueryKey, useQuery } from "@tanstack/react-query";
...
...
    const { data } = useQuery({
        queryKey: ["data", serviceScope, config],
        queryKeyHashFn: (key) => {
            const fixKeys = key.map((k) => {
                if (k instanceof ServiceScope) {
                    return serviceScope.id.toString();
                } else {
                    return k;
                }
            });
            return hashQueryKey(fixKeys);
        },
        queryFn: () => {
            return getData({ serviceScope, config  });
        },
    });
````
This way, I keep the original `hashQueryKey`, but I hook the non serializable type to extract some internal ID.

Thanks for the suggestion !
